### ISSUE TITLE
[BUG] cross_entropy_with_selfnorm

### ISSUE THREAD

**Issue by yu239** [PaddlePaddle/Paddle#1087]
Currently this cost layer assumes that the input is already softmax normalized (i.e., sum-to-one). Thus the second line of the following code is wrong:

```
Matrix::resizeOrCreate(sftMaxSum_, output.getHeight(), 1, false, useGpu_);                                                                                                                                                       
output.rowSum(*sftMaxSum_);                                                                                                                                                                                                      
sftMaxSum_->log2();          
```

Because the rowSum is always 1. I think this layer needs the non-normalized vector as the input so that it is able to sum the normalizer.

**Comment by yu239** [comment]
Sorry I understand this in a wrong way. The input can be a vector of all positive numbers, for example, after taking exp(). In this case, this cost layer makes sense. Anyway, I think the code comment should be more detailed so that the use of this layer is clarified.
